common/scheduler/scheduler_open.cc

```diff
@@ -872,17 +884,54 @@ void SchedulerOpen::setFrequency(int coreCounter, int frequency) {
 	int oldFrequency = Sim()->getMagicServer()->getFrequency(coreCounter);
 
 	if (frequency < minFrequency) {
+		cout << "[Scheduler] [Warning]: requested frequency " << frequency << " increased to minimum frequency " << minFrequency << endl;	
 		frequency = minFrequency;
 	}
-	if (frequency > maxFrequency) {
-		frequency = maxFrequency;
+	if (frequency > maxFrequencyDynamic.at(coreCounter)) {
+		cout << "[Scheduler] [Warning]: requested frequency " << frequency << " lowered to maximum frequency " << maxFrequencyDynamic.at(coreCounter) << endl;	
+		frequency = maxFrequencyDynamic.at(coreCounter);
 	}
 
 	if (frequency != oldFrequency) {
 		Sim()->getMagicServer()->setFrequency(coreCounter, frequency);
 	}
 }
 
+/** checkFrequencies
+ * Calculate maximum frequency based on change in threshold voltage and adjust core frequencies if needed.
+ */
+void SchedulerOpen::checkFrequencies() {
+	std::vector<double> vdds = performanceCounters->getVddOfCores(numberOfCores);
+	std::vector<double> delta_vs;
+	if (subcore_enabled) {
+		delta_vs = performanceCounters->getDeltaVthOfCores(numberOfComponents);
+	} else {
+		delta_vs = performanceCounters->getDeltaVthOfCores(numberOfCores);
+	}
+
+	double vdd;
+	double delta_v;
+	if (subcore_enabled) {
```

Comment:
Shouldn't this if-statement be part of the for-loop so that it sets delta_v of each core to the highest delta_v of the subcomponents of that core? Now it seems to sets delta_v to the maximum of all subcomponents of the whole cpu. And vdd also seems to be set to the global maximum. Is that correct?

Reply:
Yes. I am aware of this issue and I addressed it in my thesis. It affects my results, so it needs to stay consistent with my thesis for now, but I have added a note so that it can be fixed later.